File: source-connector/src/main/java/sourceconnector/repository/file/DecompressingFileRepository.java
package sourceconnector.repository.file;

import lombok.RequiredArgsConstructor;
import offsetmanager.domain.file.FileKey;
import sourceconnector.repository.file.decompressor.Decompressor;
import sourceconnector.repository.file.decompressor.DecompressorSelector;

import java.io.IOException;
import java.io.InputStream;

@RequiredArgsConstructor
public class DecompressingFileRepository implements FileRepository {
  private final FileRepository delegate;

  @Override
  public InputStream getFile(FileKey fileKey) throws IOException {
    InputStream inputStream = delegate.getFile(fileKey);
    Decompressor decompressor = DecompressorSelector.select(fileKey);
    return decompressor.decompress(inputStream);
  }
}

File: source-connector/src/main/kotlin/sourceconnector/repository/file/DecompressingFileRepository.kt
package sourceconnector.repository.file

import offsetmanager.domain.file.FileKey
import sourceconnector.repository.file.decompressor.DecompressorSelector.Companion.select
import java.io.IOException
import java.io.InputStream

class DecompressingFileRepository(
  private val delegate: FileRepository
) : FileRepository {

  @Throws(IOException::class)
  override fun getFile(fileKey: FileKey): InputStream {
    val inputStream = delegate.getFile(fileKey)
    val decompressor = select(fileKey)
    return decompressor.decompress(inputStream)
  }
}
